### ISSUE TITLE
[expo-location] ios doesn't support background when-in-use

### ISSUE THREAD

**Issue by jnoyola** [expo/expo#15066]
### Summary

Please correct me if I'm mistaken, but it appears that expo-location doesn't conform to the iOS spec in terms of when the "when in use" vs. "always" location permission is applicable.

From the [Apple documentation](https://developer.apple.com/documentation/corelocation/choosing_the_location_services_authorization_to_request)
* "Location services are available to apps with CLAuthorizationStatus.authorizedWhenInUse only while the app is 'in use'."
* "an app is considered in use: [...] When the app shows the background location usage indicator (showsBackgroundLocationIndicator)."

However according to the [Expo Location documentation](https://docs.expo.dev/versions/latest/sdk/location/#background-location-methods)
* The only method that supports `showsBackgroundLocationIndicator` is `Location.startLocationUpdatesAsync`.
* `Location.startLocationUpdatesAsync` is a background location method.
* "In order to use Background Location Methods [...] On iOS it must be granted with Always option"

___

These are pretty clearly not in agreement.

**According to Apple**: the "when in use" permission should allow you to use location services when the app is running in the foreground or otherwise showing the background location usage indicator.

**According to Expo**: the "always" permission is required in order to show the background location usage indicator.

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

bare

### What platform(s) does this occur on?

iOS

### SDK Version (managed workflow only)

41.0.1

### Environment

  Expo CLI 4.2.1 environment info:
    System:
      OS: macOS 11.2.2
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 14.15.0 - /usr/local/bin/node
      npm: 6.14.8 - /usr/local/bin/npm
      Watchman: 2021.08.02.00 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.10.2 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: iOS 14.5, DriverKit 20.4, macOS 11.3, tvOS 14.5, watchOS 7.4
    IDEs:
      Xcode: 12.5.1/12E507 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^41.0.1 => 41.0.1 
      react: 16.13.1 => 16.13.1 
      react-dom: 16.13.1 => 16.13.1 
      react-native: 0.63.4 => 0.63.4 
      react-native-web: ~0.13.12 => 0.13.18 
    npmGlobalPackages:
      expo-cli: 4.2.1
    Expo Workflow: bare

### Reproducible demo or steps to reproduce from a blank project

I'm fairly certain that this doesn't work by design, but based on the docs linked in the issue summary, it would appear the design is flawed.

```typescript
await Location.requestForegroundPermissionsAsync();
await Location.startLocationUpdatesAsync(
  'LOCATION_UPDATER',
  {
    accuracy: Location.LocationAccuracy.High,
    showsBackgroundLocationIndicator: true
  });
```

**Comment by jnoyola** [comment]
Essentially what I'm looking for is a way to get the ["blue pill" background location tracking](https://cdn4syt-solveyourtech.netdna-ssl.com/wp-content/uploads/2020/03/why-blue-shading-clock-iphone-1.jpg) with only the "when in use" permission.

I verified that I am able to achieve this using `react-native-geolocation-service`, but I don't believe `expo-location` can achieve this with either `startLocationUpdatesAsync` or `watchPositionAsync`.

**Comment by daku** [comment]
This is true. Our app is able to get location updates while in background even when user selects "When in use". We use `react-native-geolocation-service` which shows the blue pill to satisfy the in-use requirement. I was looking for how to do this with expo when I came across this thread.

**Comment by daku** [comment]
Noooooo! This here is me being active!

**Comment by daku** [comment]
Activity!

**Comment by johnnyrockfort** [comment]
same here

**Comment by maethub** [comment]
We are currently struggling with the same problem. Is there anyone with a solution for this? 

Falling back to a native geolocation plugin (e.g. `react-native-background-geolocation`) would mean to eject our project and leave all those expo features behind...

@daku @jnoyola did you come to a solution?

**Comment by akhzarna** [comment]
@daku is suggesting library which updated 3 years ago

**Comment by daku** [comment]
Activity

**Comment by daku** [comment]
We are still transitioning to expo and so waiting on this. We may look into
it if we have time.

**Comment by daku** [comment]
Not solved

**Comment by tknuts** [comment]
Same issue our customers get spoked by the required "Always". Other trackers can track in background with "When in use" permission only, Why not in Expo apps ?

**Comment by AlexTech314** [comment]
Has anyone found a work-around? I'm not particularly looking to eject my project. Strange how expo missed the mark on this fairly common location use case...

**Comment by RubenEngel** [comment]
Having the same issue. We want to track location while the app is minimized but not fully closed, and our delivery riders are not selecting the 'Always' setting (I probably wouldn't either)

**Comment by tknuts** [comment]
Not solved

**Comment by alfredvaa** [comment]
Seems to me that this could be resolved by defining the foregroundService property of startLocationUpdatesAsync. With that, the task manager task will be a foreground task and therefore the "always" permission is not required. I think you get the "blue pill" indication when the foregroundService property is defined as well.

I really think the docs should be more specific about this though, since the startLocationUpdatesAsync is not only for background services but for foreground services as well (which does not require the always-permission) I think most fitness tracking services are using a foreground service in this way, at least from my understanding.

In summary, a task defined by expo-task-manager can be either a background task or a foreground task. Which it is depends on if the foregroundService property on the startLocationUpdatesAsync is defined or not. A background task needs the always permission but a foreground service does not. A foreground service is killed when the app is killed, but not when the phone is locked or if another app is in focus.

**Comment by tknuts** [comment]
Hi Alfred. I think i tried that. Was and is still desperate to fix this. But I'll give it another try. It died after a short time of tracking. Could you provide me with a working sample?

**Comment by alfredvaa** [comment]
@tknuts sure!

This should be outside of any component lifecycle:

```
import * as TaskManager from "expo-task-manager";
import { LocationAccuracy, LocationObject } from "expo-location";

const TRACKING_BACKGROUND_TASK = "tracking-background-task";

const initTrackingBackgroundTask = () => {
  if (TaskManager.isTaskDefined(TRACKING_BACKGROUND_TASK)) {
    return;
  }

  TaskManager.defineTask(
    TRACKING_BACKGROUND_TASK,
    (body: TaskManager.TaskManagerTaskBody<object>) => {
      if (body.error) {
        console.error(body.error.message);
        return;
      }

      const data = body.data as unknown as { locations: LocationObject[] };

      const locations = data?.locations;

      if (!locations) {
        return;
      }

      for (const location of locations) {
        for (const callback of Object.values(locationCallbacks)) {
          callback(location);
        }
      }
    }
  );
};

initTrackingBackgroundTask();

const locationCallbacks: {
  [key: string]: React.Dispatch<
    React.SetStateAction<Location.LocationObject | null>
  >;
} = {};

function addLocationCallback(
  callback: React.Dispatch<React.SetStateAction<Location.LocationObject | null>>
) {
  const id = v4() as string;

  locationCallbacks[id] = callback;

  return {
    remove: () => {
      delete locationCallbacks[id];
    },
  };
}
```

And then within your geolocation component:

```
const Tracking = () => {
  const [location, setLocation] = React.useState<LocationObject | null>(null);

  React.useEffect(() => {
    const subscription = addLocationCallback(setLocation);

    return () => {
      subscription.remove();
    };
  }, []);

  React.useEffect(() => {
    (async () => {
      const { status: foregroundStatus } =
        await Location.requestForegroundPermissionsAsync();

      if (foregroundStatus !== "granted") {
        Alert.alert(
          "Error",
          "Turn on location services"
        );
        return;
      }

      try {
        await Location.startLocationUpdatesAsync(TRACKING_BACKGROUND_TASK, {
          distanceInterval: 1,
          accuracy: LocationAccuracy.BestForNavigation,
          foregroundService: {
            notificationTitle: "Android notification title",
            notificationBody: "Android notification body",
          },
        });
      } catch (e) {
        console.log("Error starting location updates");
        console.log(e);
      }
    })();

    return () => {
      Location.stopLocationUpdatesAsync(TRACKING_BACKGROUND_TASK);
    };
  }, []);

// Render something and use thie location variable 
return <View />;
}
```

I think that this is the important parts of what is needed to make it work, I am using Expo 48 and it seems to work there!

**Comment by zorzpapaduby** [comment]
Did this work?

**Comment by Anurag-Raut** [comment]
did the code provided by @alfredvaa solved the issue ?
